### PR TITLE
Make StreamingResponse compatible with normal iterators

### PR DESCRIPTION
**Pull request by tiangolo**
Make `StreamingResponse` compatible with normal iterators.

Continuing from #457, https://github.com/encode/starlette/pull/457#issuecomment-478965259

---

I also removed the mention to `iterator_to_async` from the docs, only mention that it supports both async and normal generators.

> I'd also be interested in if there's any nice ways around we could use something like this and drop `aiofiles` [...]

Let's see what do you think about this one and then I'll take a stab at that :smile:

### CODE REVIEW

File: starlette/concurrency.py [hunk 1]
```diff
@@ -22,3 +23,24 @@
...
+
+def _interceptable_next(iterator: Iterator) -> Any:
+    try:
+        result = next(iterator)
+        return result
+    except StopIteration:
+        raise _StopSyncIteration
+
+
+async def iterator_to_async(iterator: Iterator) -> AsyncGenerator:
```
Comment (lovelydinosaur) [inline starlette/concurrency.py]: How about `iterate_in_threadpool` to mirror `run_in_threadpool`.

Comment (tiangolo) [reply starlette/concurrency.py]: Great idea. I struggled a lot to name it and wasn't quite convinced either.

**Comment by lovelydinosaur** [comment]
Do we strictly need the `_StopSyncIteration` indirection?

**Comment by lovelydinosaur** [comment]
Can we not catch StopIteration directly?

**Comment by tiangolo** [comment]
`StopIteration` is used internally by coroutines, it was used before to yield values, here's a short summary in the PEP: https://www.python.org/dev/peps/pep-0492/#why-stopasynciteration

~I can think of an alternative using classes, based on this example: https://www.python.org/dev/peps/pep-0492/#example-2, would you prefer that?~

**Comment by tiangolo** [comment]
Because `StopIteration` was used to yield values, catching it directly in `iterate_in_threadpool` is not possible, it is converted to a `RuntimeError`.

Actually, now that I think of it, using a class would still need the indirection, as we need an exception that is propagated up/out even when run inside of `run_in_threadpool`.

**Comment by lovelydinosaur** [comment]
👍 Released as prerelease version 0.12.0b3

**Comment by tiangolo** [comment]
Awesome! Thanks! :tada: :rocket:

**Pull request closed without merging**